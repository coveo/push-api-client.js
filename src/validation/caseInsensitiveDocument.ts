export type Recordable<T> = {
  [K in keyof T]: T[K];
};

export class CaseInsensitiveDocument<T> {
  private _originalDocument: Record<string, T> = {};
  private _remainingRecord: Record<string, T> = {};

  public constructor(doc: Record<string, T>) {
    this._originalDocument = this.shallowCopy(doc);

    Object.entries(doc).forEach(([k, v]) => {
      this._remainingRecord[k.toLowerCase()] = v;
    });
  }

  public delete(...keys: string[]) {
    keys.forEach((key) => {
      delete this._remainingRecord[key];
    });
  }

  public get remainingRecord(): Readonly<Record<string, T>> {
    return this._remainingRecord;
  }

  public get originalDocument(): Readonly<Record<string, T>> {
    return this._originalDocument;
  }

  private shallowCopy(sourceRecord: Record<string, T>) {
    // No need to copy nested properties
    return Object.assign({}, sourceRecord);
  }
}
