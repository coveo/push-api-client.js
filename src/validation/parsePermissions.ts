import {PathLike} from 'fs';
import {
  ArrayValue,
  BooleanValue,
  PrimitivesValues,
  RecordValue,
  StringValue,
} from '@coveo/bueno';
import {Document} from '../document';
import {AnySecurityIdentityBuilder} from '../securityIdentityBuilder';
import {CaseInsensitiveDocument, Recordable} from './caseInsensitiveDocument';
import {DocumentBuilder} from '../documentBuilder';
import {InvalidDocument} from '../errors';
import {KnownKeys} from './knownKey';
import {PermissionSection, PermissionSetBuilder} from '../permissionSetBuilder';
import {RequiredKeyValidator} from './requiredKeyValidator';
import {
  PermissionSet,
  PermissionLevel,
  Permission,
  PermissionIdentityType,
} from '@coveo/platform-client';

export const processPermissionList = (
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentBuilder: DocumentBuilder,
  documentPath: PathLike
) => {
  ensurePermissionArray(caseInsensitiveDoc, documentPath);
  new KnownKeys<Document['permissions']>(
    'permissions',
    caseInsensitiveDoc
  ).whenExists((permissions) => {
    permissions!.forEach(
      (permission: Recordable<PermissionSet> | Recordable<PermissionLevel>) => {
        const caseInsensitivePermission =
          new CaseInsensitiveDocument<PrimitivesValues>(permission);

        new KnownKeys('permissionsets', caseInsensitivePermission)
          .whenExists<PermissionLevel>((permissionLevel) =>
            processPermissionLevel(
              permissionLevel,
              documentBuilder,
              documentPath
            )
          )
          .whenDoesNotExist<PermissionSet>((permissionSet) =>
            processPermissionSet(permissionSet, documentBuilder, documentPath)
          );
      }
    );

    delete caseInsensitiveDoc.documentRecord['permissions'];
  });
};

const processPermissionSet = (
  permissionSet: PermissionSet,
  documentBuilder: DocumentBuilder,
  documentPath: PathLike
) => {
  const permissionSetBuilder =
    validateRequiredPermissionSetKeysAndGetPermissionSetBuilder(
      permissionSet,
      documentPath
    );

  documentBuilder.withPermissionSet(permissionSetBuilder);
};

const processPermissionLevel = (
  permission: Recordable<PermissionLevel>,
  documentBuilder: DocumentBuilder,
  documentPath: PathLike
) => {
  // TODO: CDX-1282: require at least some permissions
  const permissionSetBuilders = permission.permissionSets!.map(
    (permissionSet) => {
      const caseInsensitivePermissions = new CaseInsensitiveDocument(
        permission
      );
      const requiredPermissionLevelName = new RequiredKeyValidator<string>(
        'name',
        caseInsensitivePermissions,
        new StringValue({required: true, emptyAllowed: false})
      );
      if (!requiredPermissionLevelName.isValid) {
        throw new InvalidDocument(
          documentPath,
          requiredPermissionLevelName.explanation
        );
      }
      return validateRequiredPermissionSetKeysAndGetPermissionSetBuilder(
        permissionSet,
        documentPath
      );
    }
  );

  // TODO: CDX-1282: prevent empty permission set
  documentBuilder.withPermissionLevel(permission.name!, permissionSetBuilders);
};

const validateAllowAnonymous = (
  caseInsensitivePermissions: CaseInsensitiveDocument<PrimitivesValues>,
  documentPath: PathLike
): boolean | never => {
  const requiredAllowAnonymous = new RequiredKeyValidator(
    'allowanonymous',
    caseInsensitivePermissions,
    new BooleanValue({required: true})
  );
  if (!requiredAllowAnonymous.isValid) {
    throw new InvalidDocument(documentPath, requiredAllowAnonymous.explanation);
  }
  return requiredAllowAnonymous.value!;
};

const validatePermission = (
  permissionLogic: Lowercase<PermissionSection>,
  caseInsensitivePermissions: CaseInsensitiveDocument<PrimitivesValues>,
  documentPath: PathLike
) => {
  const requiredPermission = new RequiredKeyValidator(
    permissionLogic,
    caseInsensitivePermissions,
    getSecurityIdentitySchemaValidation()
  );

  if (!requiredPermission.isValid) {
    throw new InvalidDocument(documentPath, requiredPermission.explanation);
  }

  return requiredPermission.value;
};

const validateRequiredPermissionSetKeysAndGetPermissionSetBuilder = (
  permission: Recordable<PermissionSet>,
  documentPath: PathLike
): PermissionSetBuilder => {
  const caseInsensitivePermissions = new CaseInsensitiveDocument(permission);
  const allowAnonymous = validateAllowAnonymous(
    caseInsensitivePermissions,
    documentPath
  );
  const allowedPermissions = validatePermission(
    'allowedpermissions',
    caseInsensitivePermissions,
    documentPath
  );
  const deniedPermissions = validatePermission(
    'deniedpermissions',
    caseInsensitivePermissions,
    documentPath
  );

  const permissionSetBuilder = new PermissionSetBuilder(allowAnonymous);

  allowedPermissions?.forEach(
    ({identity, identityType, securityProvider}: Permission) => {
      permissionSetBuilder.withAllowedPermissions(
        new AnySecurityIdentityBuilder(
          identityType!,
          identity!,
          securityProvider
        )
      );
    }
  );

  deniedPermissions?.forEach(
    ({identity, identityType, securityProvider}: Permission) => {
      permissionSetBuilder.withDeniedPermissions(
        new AnySecurityIdentityBuilder(
          identityType!,
          identity!,
          securityProvider
        )
      );
    }
  );

  return permissionSetBuilder;
};

const getIdentityTypeRegex = () => {
  const identityTypesValues = Object.values(PermissionIdentityType);
  return new RegExp(identityTypesValues.join('|'), 'i');
};

const getSecurityIdentitySchemaValidation = (): ArrayValue<Permission> => {
  return new ArrayValue({
    required: false,
    each: new RecordValue({
      values: {
        identity: new StringValue({required: true, emptyAllowed: false}),
        identityType: new StringValue({
          regex: getIdentityTypeRegex(),
          required: true,
          emptyAllowed: false,
        }),
        securityProvider: new StringValue({
          emptyAllowed: true,
          required: false,
        }),
      },
    }),
  });
};

const ensurePermissionArray = (
  caseInsensitiveDoc: CaseInsensitiveDocument<PrimitivesValues>,
  documentPath: PathLike
) => {
  const validator = (schema: ArrayValue<PrimitivesValues>) =>
    new RequiredKeyValidator('permissions', caseInsensitiveDoc, schema);
  const requiredPermissionArray = validator(new ArrayValue({required: false}));
  const requiredAtLeastOnePermission = validator(new ArrayValue({min: 1}));

  if (!requiredPermissionArray.isValid) {
    throw new InvalidDocument(
      documentPath,
      requiredPermissionArray.explanation
    );
  }

  if (!requiredAtLeastOnePermission.isValid) {
    // simply discard empty array permission instead of throwing
    delete caseInsensitiveDoc.documentRecord['permissions'];
  }
};
