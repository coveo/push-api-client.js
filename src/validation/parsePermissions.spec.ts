import {PrimitivesValues} from '@coveo/bueno';
import {join} from 'path';
import {DocumentBuilder} from '..';
import {CaseInsensitiveDocument} from './caseInsensitiveDocument';
import {processPermissionList} from './parsePermissions';

describe('parsePermissions', () => {
  let documentBuilder: DocumentBuilder;
  const dummyPath = join('dummy', 'path');
  const publicPermissionsSet = {
    allowAnonymous: true,
    deniedpermissions: [
      {
        identity: 'foo@example.com',
        identityType: 'USER',
      },
    ],
  };
  const restrictedPermissionSet = {
    allowAnonymous: false,
    ALLOWEDPERMISSIONS: [
      {
        identity: 'asmith@example.com',
        identityType: 'USER',
      },
      {
        identity: 'bjones@example.com',
        identityType: 'USER',
      },
    ],
  };

  const complexPermissionSet = {
    aLLowAnonyMOUS: false,
    deniedpermissions: [
      {
        identity: 'foo',
        identityType: 'VIRTUAL_GROUP',
      },
    ],
    allowedPERMISSIONS: [
      {
        identity: 'Sample_group',
        identityType: 'GROUP',
      },
      {
        identity: 'bjones@example.com',
        identityType: 'USER',
      },
    ],
  };

  beforeEach(() => {
    documentBuilder = new DocumentBuilder('https://foo.com', 'Some Document');
  });

  it('should not marhsal empty permssion when is empty array', () => {
    const caseInsensitiveDoc = new CaseInsensitiveDocument<PrimitivesValues>({
      permissions: [],
    });
    processPermissionList(caseInsensitiveDoc, documentBuilder, dummyPath);
    expect(documentBuilder.marshal()).not.toHaveProperty('permissions');
  });

  it('should marhsal denied permssion set', () => {
    const caseInsensitiveDoc = new CaseInsensitiveDocument<PrimitivesValues>({
      permissions: [publicPermissionsSet],
    });
    processPermissionList(caseInsensitiveDoc, documentBuilder, dummyPath);
    expect(documentBuilder.marshal().permissions).toMatchSnapshot();
  });

  it('should marhsal allowed permssion set', () => {
    const caseInsensitiveDoc = new CaseInsensitiveDocument<PrimitivesValues>({
      permissions: [restrictedPermissionSet],
    });
    processPermissionList(caseInsensitiveDoc, documentBuilder, dummyPath);
    expect(documentBuilder.marshal().permissions).toMatchSnapshot();
  });

  it('should marhsal combined permssion sets', () => {
    const caseInsensitiveDoc = new CaseInsensitiveDocument<PrimitivesValues>({
      permissions: [
        publicPermissionsSet,
        restrictedPermissionSet,
        complexPermissionSet,
      ],
    });
    processPermissionList(caseInsensitiveDoc, documentBuilder, dummyPath);
    expect(documentBuilder.marshal().permissions).toMatchSnapshot();
  });
});
