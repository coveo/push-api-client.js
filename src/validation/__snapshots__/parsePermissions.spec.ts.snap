// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePermissions should marhsal allowed permssion set 1`] = `
[
  {
    "allowAnonymous": false,
    "allowedPermissions": [
      {
        "identity": "asmith@example.com",
        "identityType": "User",
      },
      {
        "identity": "bjones@example.com",
        "identityType": "USER",
      },
    ],
    "deniedPermissions": [],
  },
]
`;

exports[`parsePermissions should marhsal combined permssion sets 1`] = `
[
  {
    "allowAnonymous": true,
    "allowedPermissions": [],
    "deniedPermissions": [
      {
        "identity": "foo@example.com",
        "identityType": "USER",
      },
    ],
  },
  {
    "allowAnonymous": false,
    "allowedPermissions": [
      {
        "identity": "asmith@example.com",
        "identityType": "User",
      },
      {
        "identity": "bjones@example.com",
        "identityType": "USER",
      },
    ],
    "deniedPermissions": [],
  },
  {
    "allowAnonymous": false,
    "allowedPermissions": [
      {
        "identity": "Sample_group",
        "identityType": "Group",
      },
      {
        "identity": "bjones@example.com",
        "identityType": "user",
      },
    ],
    "deniedPermissions": [
      {
        "identity": "foo",
        "identityType": "VirtualGroup",
      },
    ],
  },
]
`;

exports[`parsePermissions should marhsal denied permssion set 1`] = `
[
  {
    "allowAnonymous": true,
    "allowedPermissions": [],
    "deniedPermissions": [
      {
        "identity": "foo@example.com",
        "identityType": "USER",
      },
    ],
  },
]
`;
